src/utilities/validators/base-validator.ts

```diff
@@ -11,33 +11,54 @@ import {
   TypeMismatch,
   ValidationFailure,
 } from '../../validation-messages/failures';
+import ValidationMessage from '../../validation-messages/validation-message';
 import {
   EmptyArray,
   MissingProperties,
   ValidationWarning,
 } from '../../validation-messages/warnings';
 
 abstract class BaseValidator {
-  protected _failures: ValidationFailure[];
+  protected _failures: Map<string, ValidationFailure>;
 
-  protected _warnings: ValidationWarning[];
+  protected _warnings: Map<string, ValidationWarning>;
 
   protected validated: boolean;
 
   constructor() {
     this.validated = false;
-    this._failures = [];
-    this._warnings = [];
+    this._failures = new Map();
+    this._warnings = new Map();
   }
 
-  public get failures(): ValidationFailure[] {
+  public get failures(): Map<string, ValidationFailure> {
     return this._failures;
   }
 
-  public get warnings(): ValidationWarning[] {
+  public get warnings(): Map<string, ValidationWarning> {
     return this._warnings;
   }
 
+  protected addFailure(failure: ValidationFailure): void {
+    this.addValidationMessage(failure, this._failures);
```

Comment:
I agree, there are more improvements, but this definitely going to help.  In further discussions, we could further decouple this code and create the ValidationMessage/Map<string, ValidationMessage> object in a different class that is brought in here instead of defining it in this code.  Might prove to be more useful further down the road as a more generic container.